lib/ansible/modules/cloud/vmware/vmware_vswitch.py

```diff
@@ -195,179 +354,428 @@ def process_state(self):
+        vss_spec.policy.nicTeaming.failureCriteria.checkErrorPercent = False
+        vss_spec.policy.nicTeaming.failureCriteria.checkDuplex = False
+        vss_spec.policy.nicTeaming.failureCriteria.speed = 10
+        vss_spec.policy.nicTeaming.failureCriteria.checkSpeed = 'minimum'
+        if self.module.check_mode:
+            results['changed'] = True
+            results['result'] = "vSwitch '%s' will be created" % self.switch
+        else:
+            try:
+                network_mgr = self.host_system.configManager.networkSystem
+                if network_mgr:
+                    network_mgr.AddVirtualSwitch(vswitchName=self.switch,
+                                                 spec=vss_spec)
+                    results['changed'] = True
+                    results['result'] = "vSwitch '%s' was created successfully" % self.switch
+                else:
+                    self.module.fail_json(msg="Failed to find network manager for ESXi system")
+            except vim.fault.AlreadyExists as already_exists:
+                results['result'] = "vSwitch with name %s already exists: %s" % (self.switch,
+                                                                                 to_native(already_exists.msg))
+            except vim.fault.ResourceInUse as resource_used:
+                self.module.fail_json(msg="Failed to add vSwitch '%s' as physical network adapter"
+                                          " being bridged is already in use: %s" % (self.switch,
+                                                                                    to_native(resource_used.msg)))
+            except vim.fault.HostConfigFault as host_config_fault:
+                self.module.fail_json(msg="Failed to add vSwitch '%s' due to host"
+                                          " configuration fault : %s" % (self.switch,
+                                                                         to_native(host_config_fault.msg)))
+            except vmodl.fault.InvalidArgument as invalid_argument:
+                self.module.fail_json(msg="Failed to add vSwitch '%s', this can be due to either of following :"
+                                          " 1. vSwitch Name exceeds the maximum allowed length,"
+                                          " 2. Number of ports specified falls out of valid range,"
+                                          " 3. Network policy is invalid,"
+                                          " 4. Beacon configuration is invalid : %s" % (self.switch,
+                                                                                        to_native(invalid_argument.msg)))
+            except vmodl.fault.SystemError as system_error:
+                self.module.fail_json(msg="Failed to add vSwitch '%s' due to : %s" % (self.switch,
+                                                                                      to_native(system_error.msg)))
+            # TODO: delete
```

Comment:
I would say keep the generic exception rather than removing them altogether,

Reply:
If you check the documentation, generic errors never happen for any Pyvmoni function. It's cleaner to remove those generic exceptions. It's also best practice to not use generic exceptions.

Reply:
@ckotte you're right about wholesale exceptions in general. However, there's one case when it's okay: on the top level layer of the app (in this case the module).
The thing is that if the process will just crash it won't return a failed json struct to the Ansible controller, which won't be able to handle this better then. So I'd put such handler around (some part of?) main func.

Reply:
@webknjaz Does it makes sense to do it like this in the main function?
```
try:
  host_virtual_switch = VMwareHostVirtualSwitch(module)
  host_virtual_switch.process_state()
except Exception as e:
  module.fail_json(msg=to_native(e))
```

Reply:
Yeah, smth like that. Input args might also be useful there. Basically any details you're able to retrieve is good :)

Reply:
Do you mean putting `argument_spec` in the error message as well? Could you point me to code from another module that has a similar "feature" already implemented?

Reply:
Sorry, I cannot point to one. It's just an idea though, not a strict rule or requirement.